src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java

```diff
@@ -751,20 +779,42 @@ public Node call() throws Exception {
                                                 String.format("Failure to in post-provisioning for '%s'", vmName),
                                                 e);
 
-                                        // Attempt to terminate whatever was created
+                                        handleFailure(template, vmName, e, FailureStage.POSTPROVISIONING);
+
+                                        // Remove the node from jenkins
+                                        try {
+                                            Jenkins.getInstance().removeNode(agent);
+                                        } catch (IOException nodeRemoveEx) {
+                                            LOGGER.log(
+                                                    Level.SEVERE,
+                                                    String.format("Failure removing Jenkins node for '%s'", vmName),
+                                                    nodeRemoveEx);
+                                            // Do not throw to avoid it being recorded
+                                        }
+
+                                        throw AzureCloudException.create(e);
```

Comment:
I made a bit changes. Now the method signature is `call() throws AzureCloudException` and here we convert any exception to `AzureCloudException` to meet this restriction.